src/nspawn/nspawn.c

```diff
@@ -3206,6 +3206,17 @@ static int outer_child(
         return 0;
 }
 
+static int on_sigchld(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata) {
+        pid_t pid;
+
+        pid = PTR_TO_UINT32(userdata);
+        if (pid == si->ssi_pid) {
```

Comment:
Oh, indeed, `--share-system` (and `SYSTEMD_NSPAWN_SHARE_NS_PID=yes`) causes that problem.

> why exactly kernel reparents it to nspawn and not to the init(pid=1) as it should normally do

`systemd-nspawn` calls [`prctl(PR_SET_CHILD_SUBREAPER, 1)`](https://github.com/systemd/systemd/blob/5367354dae90f0901ec3b96cbbc9f96bf4a71d11/src/nspawn/nspawn.c#L4312)